Sources/AsyncHTTPClient/HTTPHandler.swift

```diff
@@ -901,6 +899,19 @@ extension TaskHandler: ChannelDuplexHandler {
         }
     }
 
+    private func writeBodyPart(context: ChannelHandlerContext, part: IOData, promise: EventLoopPromise<Void>) {
+        switch self.state {
+        case .idle:
+            self.actualBodyLength += part.readableBytes
```

Comment:
I _think_ (would need a test here too) that we're policing the body length too late here, no? We seem to just add the readableBytes but I can't see that we check that it's actually less than or equal the number of body bytes allowed.

What happens if we do say `content-length: 1` and then send `XABCDEFG`, aren't we then still sending `ABCDEFG`? I think we are and that'd be a security vulnerability because we could smuggle a request (say we did `XGET /something HTTP/1.1\r\n\r\n`

Reply:
thats a great point, yeah, we only check in the end, let me see how I can handle it here

Reply:
ok, I added an early fail here

Reply:
@artemredkin this _looks_ good but given that it's security relevant, could we add a test specifically for that case (one that'd fail without the new condition)

Reply:
we already have it, by accident, we have two tests for content-length/body-length checks:
 - testContentLengthTooLongFails
 - testContentLengthTooShortFails
Last one will fail at the end of the request, before we send `.end`, but the first one will now be failed by this new check

Reply:
Potentially we can split the error message, just to be sure

Reply:
@artemredkin I think we need a new one that tests that we do _not_ send the bytes.

Reply:
Because as soon as the bytes are on the wire, we created a security vulnerability. Even if we detect later that something's wrong, it'll be too late.

Reply:
@Lukasa same here, I've added a test to check that we do not send more bytes then body length, although I'm not 100% sure its foolproof